app/services/tunnel.py

```diff
@@ -131,26 +144,37 @@ def get_tunnel_details(self, result: dict, job_id: str) -> Tuple[str, str]:
         ip_address = next(x["Address"] for x in nodes if x["ID"] in allocation_node)
         allocated_ports = values(allocation_info, "Resources/Networks/0/DynamicPorts/*")
         ssh_port = next(x for x in allocated_ports if x["Label"] == "ssh")["Value"]
-
         if current_app.config["ENV"] == "development":
             ip_address = current_app.config["SEA_HOST"]
 
         return (ssh_port, ip_address)
 
+    def get_tcp_port(self) -> int:
+        try:
+            new_port = redis_client.spop("open_tcp_ports")
+            return int(new_port)
+        except redis.exceptions.ConnectionError:
+            raise RedisError(detail="Redis is down")
+        except:
+            raise TunnelError(detail="No tcp ports available")
+
 
 class TunnelDeletionService:
     def __init__(self, current_user: User, tunnel: Optional[Tunnel], job_id=None):
         self.current_user = current_user
         self.tunnel = tunnel
         if job_id:
             self.job_id = job_id
-        if tunnel:
+        if self.tunnel:
             self.subdomain = tunnel.subdomain
             self.job_id = tunnel.job_id
 
         self.nomad_client = nomad.Nomad(discover_service("nomad").ip)
 
     def delete(self):
+        if self.tunnel.tcp_ports:
+            for port in self.tunnel.tcp_ports:
+                redis_client.sadd("open_tcp_ports", port)
```

Comment:
can you do this as a set merge operation instead?  We're potentially making multiple round trips to Redis here.